# Desafío:

1. ¿Cuántas visitas se realizaron en total?
2. ¿Cuál es el promedio de propiedades por usuario?
3. ¿Cuál era la temperatura promedio de todas las visitas que realizó el usuario con ID 2?
4. ¿Cuál es la temperatura promedio de las visitas para los días con lluvia?
5. ¿Cuál es la temperatura promedio para las visitas realizadas en la localidad de Suba?


In [2]:
import pandas as pd
from datetime import datetime
import pytz

------------

## Initial data preparation

### Visits Data

In [3]:
visits_data = pd.read_csv('./data/visits.csv') 
visits_data.head(5)

scheduled_id  property_id                 begin_date  \
0        169548            1  2022-01-13T10:00:00-03:00   
1        184763            1  2022-01-26T18:00:00-03:00   
2        186092            1  2022-01-28T12:00:00-03:00   
3        182497            1  2022-01-23T11:00:00-03:00   
4        182396            1  2022-01-23T10:00:00-03:00   

                    end_date type_visit     status  
0  2022-01-13T12:00:00-03:00      Visit  Cancelled  
1  2022-01-26T20:00:00-03:00      Visit  Cancelled  
2  2022-01-28T14:00:00-03:00      Visit  Cancelled  
3  2022-01-23T13:00:00-03:00      Visit  Cancelled  
4  2022-01-23T12:00:00-03:00      Visit  Cancelled

In [4]:
visits_data.describe()

scheduled_id  property_id
count     425.000000   425.000000
mean   188094.894118    40.087059
std     13820.402675    22.943521
min    160130.000000     1.000000
25%    176671.000000    21.000000
50%    190273.000000    38.000000
75%    200307.000000    61.000000
max    208926.000000    80.000000

In [5]:
visits_data.isna().sum()

scheduled_id    0
property_id     0
begin_date      0
end_date        0
type_visit      0
status          0
dtype: int64

In [6]:
visits_data.dtypes

scheduled_id     int64
property_id      int64
begin_date      object
end_date        object
type_visit      object
status          object
dtype: object

In [7]:
# Convert datetime columns into datetime objects
visits_data['begin_date'] = pd.to_datetime(visits_data['begin_date'])
visits_data['end_date'] = pd.to_datetime(visits_data['end_date'])

visits_data.dtypes

scheduled_id                                     int64
property_id                                      int64
begin_date      datetime64[ns, pytz.FixedOffset(-180)]
end_date        datetime64[ns, pytz.FixedOffset(-180)]
type_visit                                      object
status                                          object
dtype: object

In [8]:
visits_data.iloc[0].begin_date

Timestamp('2022-01-13 10:00:00-0300', tz='pytz.FixedOffset(-180)')

In [9]:
if len(visits_data.scheduled_id.unique()) == len(visits_data):
    print('All scheduled_id are unique')

All scheduled_id are unique


### Properties Data

In [10]:
properties_data = pd.read_csv('./data/properties.csv')
properties_data.head()

property_id    type_house  business_type  bedrooms  bathrooms  \
0            1  departamento         Rental         1          1   
1            2  departamento  Rental & Sale         1          1   
2            3  departamento         Rental         2          2   
3            4  departamento  Rental & Sale         3          2   
4            5  departamento  Rental & Sale         2          3   

   parking_lots  services  balcony   pool  latitude  longitude    localidad  \
0             1       3.0        0  False  4.870956 -74.058040        Chí­A   
1             2       NaN        1  False  4.623068 -74.074030  Teusaquillo   
2             2       NaN        0  False  4.723909 -74.042336      Usaquen   
3             1       0.0        0  False  4.721716 -74.057976         Suba   
4             2       0.0        1  False  4.688163 -74.043300      Usaquen   

           city                  region   country  
0  Cundinamarca  Región De Cundinamarca  Colombia  
1        Bogotá  Región De Cundinamarca  Colombia  
2        Bogotá  Región De Cundinamarca  Colombia  
3        Bogotá  Región De Cundinamarca  Colombia  
4        Bogotá  Región De Cundinamarca  Colombia

In [11]:
properties_data.describe()

property_id   bedrooms  bathrooms  parking_lots       services  \
count      80.0000  80.000000  80.000000     80.000000      19.000000   
mean       40.5000   2.275000   1.725000      0.650000    5264.526316   
std        23.2379   0.871126   0.655551      0.713345   22941.242048   
min         1.0000   1.000000   1.000000      0.000000       0.000000   
25%        20.7500   2.000000   1.000000      0.000000       0.000000   
50%        40.5000   2.000000   2.000000      1.000000       1.000000   
75%        60.2500   3.000000   2.000000      1.000000       3.000000   
max        80.0000   5.000000   4.000000      2.000000  100000.000000   

         balcony   latitude  longitude  
count  80.000000  80.000000  80.000000  
mean    0.137500   4.685079 -74.116246  
std     0.346547   0.062736   0.062496  
min     0.000000   4.586188 -74.245544  
25%     0.000000   4.644076 -74.158101  
50%     0.000000   4.685236 -74.111243  
75%     0.000000   4.722041 -74.057694  
max     1.000000   4.926684 -74.023470

In [12]:
properties_data.isna().sum()

property_id       0
type_house        0
business_type     0
bedrooms          0
bathrooms         0
parking_lots      0
services         61
balcony           0
pool              0
latitude          0
longitude         0
localidad         0
city              0
region            0
country           0
dtype: int64

Is there a difference between services Nulls and services equal to 0?

In [13]:
properties_data.head(1)

property_id    type_house business_type  bedrooms  bathrooms  parking_lots  \
0            1  departamento        Rental         1          1             1   

   services  balcony   pool  latitude  longitude localidad          city  \
0       3.0        0  False  4.870956  -74.05804     Chí­A  Cundinamarca   

                   region   country  
0  Región De Cundinamarca  Colombia

In [14]:
properties_data.dtypes

property_id        int64
type_house        object
business_type     object
bedrooms           int64
bathrooms          int64
parking_lots       int64
services         float64
balcony            int64
pool                bool
latitude         float64
longitude        float64
localidad         object
city              object
region            object
country           object
dtype: object

### Users Data

In [15]:
users_data = pd.read_csv('./data/users.csv')
users_data.head(5)

property_id  user_id first_name last_name               address
0           31        1  Josephine   Darakjy   4 B Blue Ridge Blvd
1           34        2        Art    Venere  8 W Cerritos Ave #54
2           48        3      Lenna  Paprocki           639 Main St
3           27        4    Donette    Foller          34 Center St
4           40        5     Simona   Morasca          3 Mcauley Dr

In [16]:
users_data.describe()

property_id  user_id
count      80.0000  80.0000
mean       40.5000  40.5000
std        23.2379  23.2379
min         1.0000   1.0000
25%        20.7500  20.7500
50%        40.5000  40.5000
75%        60.2500  60.2500
max        80.0000  80.0000

In [17]:
users_data.isna().sum()

property_id    0
user_id        0
first_name     0
last_name      0
address        0
dtype: int64

In [18]:
users_data.dtypes

property_id     int64
user_id         int64
first_name     object
last_name      object
address        object
dtype: object

---------------

## Q1. ¿Cuántas visitas se realizaron en total?

In [19]:
visits_data.head(2)

scheduled_id  property_id                begin_date  \
0        169548            1 2022-01-13 10:00:00-03:00   
1        184763            1 2022-01-26 18:00:00-03:00   

                   end_date type_visit     status  
0 2022-01-13 12:00:00-03:00      Visit  Cancelled  
1 2022-01-26 20:00:00-03:00      Visit  Cancelled

In [20]:
visits_data.shape

(425, 6)

In [21]:
visits_data.status.value_counts()

Cancelled    286
Done         139
Name: status, dtype: int64

In [22]:
# maybe there is more than 1 scheduled_id per status?
visits_data.groupby('scheduled_id').count().max()

property_id    1
begin_date     1
end_date       1
type_visit     1
status         1
dtype: int64

### Solution Q1
Según la base entregada, se hicieron **139 visitas** de un total de 425 visitas agendadas.

## Q2. ¿Cuál es el promedio de propiedades por usuario?

In [23]:
users_data.head(2)

property_id  user_id first_name last_name               address
0           31        1  Josephine   Darakjy   4 B Blue Ridge Blvd
1           34        2        Art    Venere  8 W Cerritos Ave #54

In [115]:
# There should not be more than 1 user_id per property_id
users_data.groupby('property_id').count().max().user_id

1

In [116]:
# calculating mean properties per user
users_data.groupby('user_id').count().mean()

property_id    1.0
first_name     1.0
last_name      1.0
address        1.0
dtype: float64

mmm... curioso...1 propiedad por usuario

In [26]:
len(users_data.user_id.unique())

80

In [27]:
len(users_data.property_id.unique())

80

In [28]:
users_data.shape[0]

80

### Solution Q2
El **promedio de propiedades por usuario es de 1**. Es más, según la data entregada, cada usuario tiene solo 1 propierdad hasta ahora.

------

## Extracción de datos de temperatura
For the extraction of weather information per location, will use the API mentioned in the task documentation. You can obtain more info in the oficial documentation of the API: [Documentation weather API](https://www.visualcrossing.com/resources/documentation/weather-api/timeline-weather-api/) 

API needs: 
- Latitude
- Longitude
- Start Date
- End Date

An example of request: https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/{lat},{lng}/{start_date}/{end_date}?key={API_KEY}&include=days

To considered: the free tier of the API is limited to maximum of 1.000 records/day.
> Records/day - Free plans include a limit on the number of Queries Records that can be retrieved in any 24 hour period. Note that A single 15-day forecast query counts as a single Query Result while each history record returned counts as a Query Result.

So it will not be posible to get the data for each hour and day. Will get only daily weather and make analysis with that.

An option would be to assume that the temperature is the same for each hour in a day, or that the temperature is distributed normaly across the day (we have the min, max temperature per day, and the average).



Logic for implementation: 
- Each property (defined as lat + long) has a list of visits. 
- Will take the max & min visiting days of each property and get the weather for those dates ranges.
- Join the weather data with the visits data.

Main doubts:
- Does the API return different weather data for the same city but different locations? (ex: Santiago de chile, Las Condes vs. Providencia for the same day).
    <br>If not, can pick random lat+lng for each city and make a request for each city only.
    <br>**For now will asume that it's different**. But if the API requests are getting too expensive, this can be a solution.



In [29]:
properties_data.head(2)

property_id    type_house  business_type  bedrooms  bathrooms  \
0            1  departamento         Rental         1          1   
1            2  departamento  Rental & Sale         1          1   

   parking_lots  services  balcony   pool  latitude  longitude    localidad  \
0             1       3.0        0  False  4.870956  -74.05804        Chí­A   
1             2       NaN        1  False  4.623068  -74.07403  Teusaquillo   

           city                  region   country  
0  Cundinamarca  Región De Cundinamarca  Colombia  
1        Bogotá  Región De Cundinamarca  Colombia

In [30]:
visits_data.head(2)

scheduled_id  property_id                begin_date  \
0        169548            1 2022-01-13 10:00:00-03:00   
1        184763            1 2022-01-26 18:00:00-03:00   

                   end_date type_visit     status  
0 2022-01-13 12:00:00-03:00      Visit  Cancelled  
1 2022-01-26 20:00:00-03:00      Visit  Cancelled

In [31]:
# Get example for logic that has min 2 visits
top_visited_properties = visits_data[['property_id', 'scheduled_id']].groupby('property_id').count().sort_values(by='scheduled_id', ascending=False).head(1).reset_index().iloc[0]
top_visited_properties

property_id      3
scheduled_id    10
Name: 0, dtype: int64

In [32]:
test_property = properties_data[properties_data.property_id == top_visited_properties.property_id]
test_property

property_id    type_house business_type  bedrooms  bathrooms  parking_lots  \
2            3  departamento        Rental         2          2             2   

   services  balcony   pool  latitude  longitude localidad    city  \
2       NaN        0  False  4.723909 -74.042336   Usaquen  Bogotá   

                   region   country  
2  Región De Cundinamarca  Colombia

In [33]:
test_property = test_property.iloc[0]

In [34]:
min_date = visits_data.loc[visits_data['property_id'] == test_property.property_id, 'begin_date'].min()
max_date = visits_data.loc[visits_data['property_id'] == test_property.property_id, 'begin_date'].max()
print(f"min_date: {min_date}, max_date: {max_date}")

min_date: 2022-01-05 13:00:00-03:00, max_date: 2022-01-16 19:00:00-03:00


### Testing logic for `test_property`

In [35]:
import requests
import os

In [35]:
if not os.getenv('WEATHER_API_KEY', None):
    raise ValueError('WEATHER_API_KEY environment variable not set')

In [36]:
BASE_URL = "https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/{lat},{lng}/{start_date}/{end_date}?key={API_KEY}&include=days"

In [37]:
test_weather_url = BASE_URL.format(lat=test_property.latitude, lng=test_property.longitude, start_date=f"{min_date:%Y-%m-%d}", end_date=f"{max_date:%Y-%m-%d}", API_KEY=os.getenv('WEATHER_API_KEY'))
r = requests.get(url=test_weather_url)
res = r.json()

r.status_code


200

In [39]:
res

{'queryCost': 12,
 'latitude': 4.7239094,
 'longitude': -74.042336,
 'resolvedAddress': '4.7239094,-74.042336',
 'address': '4.7239094,-74.042336',
 'timezone': 'America/Bogota',
 'tzoffset': -5.0,
 'days': [{'datetime': '2022-01-05',
   'datetimeEpoch': 1641358800,
   'tempmax': 71.8,
   'tempmin': 50.8,
   'temp': 59.4,
   'feelslikemax': 71.8,
   'feelslikemin': 50.8,
   'feelslike': 59.4,
   'dew': 49.4,
   'humidity': 72.99,
   'precip': 0.0,
   'precipprob': None,
   'precipcover': 0.0,
   'preciptype': None,
   'snow': None,
   'snowdepth': None,
   'windgust': None,
   'windspeed': 13.8,
   'winddir': 192.6,
   'pressure': 1023.3,
   'cloudcover': 56.5,
   'visibility': 6.2,
   'solarradiation': None,
   'solarenergy': None,
   'uvindex': None,
   'sunrise': '06:05:56',
   'sunriseEpoch': 1641380756,
   'sunset': '17:57:27',
   'sunsetEpoch': 1641423447,
   'moonphase': 0.05,
   'conditions': 'Partially cloudy',
   'description': 'Partly cloudy throughout the day.',
   'icon': 

All the questions Aim for dates temperatures. So data needed from the API:
- datetime
- temperature

Will also get other data because I find it useful and the free tier of the API is limited. So have to make it worth it.

In [37]:
import copy

In [42]:
# will drop stations because its has more than 1 record and it will duplicate records in the dataframe. 
# also for now is not necessary so im not going to overcomplicate the code.
mask = copy.deepcopy(res["days"][0])
mask.pop("stations") # wont needed and will duplicate rows when converting to DF
pd.DataFrame([mask])

datetime  datetimeEpoch  tempmax  tempmin  temp  feelslikemax  \
0  2022-01-05     1641358800     71.8     50.8  59.4          71.8   

   feelslikemin  feelslike   dew  humidity  ...  uvindex   sunrise  \
0          50.8       59.4  49.4     72.99  ...     None  06:05:56   

   sunriseEpoch    sunset sunsetEpoch moonphase        conditions  \
0    1641380756  17:57:27  1641423447      0.05  Partially cloudy   

                         description               icon  source  
0  Partly cloudy throughout the day.  partly-cloudy-day     obs  

[1 rows x 34 columns]

In [43]:
# Save example for easier reading... and later documentation
with open('./data/examples/test_weather.json', 'w') as f:
    f.write(r.text)

For now, will make a request for each property (lat+lng). How many are they?

In [38]:
len(visits_data.property_id.unique())

80

so... will make 80 requests in total. And the "Query cost" for the API (amount of days) would be:

In [39]:
properties_begin_dates = visits_data[['property_id', 'begin_date']].groupby('property_id').min().reset_index()
properties_end_dates = visits_data[['property_id', 'end_date']].groupby('property_id').max().reset_index()
properties_visit_dates_range = pd.merge(properties_begin_dates, properties_end_dates, on='property_id')
properties_visit_dates_range.head(2)

property_id                begin_date                  end_date
0            1 2022-01-10 12:00:00-03:00 2022-01-28 14:00:00-03:00
1            2 2022-01-20 13:00:00-03:00 2022-01-20 15:00:00-03:00

In [40]:
properties_visit_dates_range.loc[properties_visit_dates_range.property_id==top_visited_properties.property_id]

property_id                begin_date                  end_date
2            3 2022-01-05 13:00:00-03:00 2022-01-16 21:00:00-03:00

In [41]:
properties_visit_dates_range.dtypes

property_id                                     int64
begin_date     datetime64[ns, pytz.FixedOffset(-180)]
end_date       datetime64[ns, pytz.FixedOffset(-180)]
dtype: object

In [42]:
properties_visit_dates_range['dates_diff'] = properties_visit_dates_range.end_date - properties_visit_dates_range.begin_date
properties_visit_dates_range.head(2)

property_id                begin_date                  end_date  \
0            1 2022-01-10 12:00:00-03:00 2022-01-28 14:00:00-03:00   
1            2 2022-01-20 13:00:00-03:00 2022-01-20 15:00:00-03:00   

        dates_diff  
0 18 days 02:00:00  
1  0 days 02:00:00

In [43]:
properties_visit_dates_range['dates_diff'].sum()

Timedelta('652 days 01:05:00')

Aprox "query cost" would be 652 (this add hours also, so it's should be less...)

And free tier API usage allow maximum 1.000 requests per day. So we should be OK.


### Getting all the data

Will get the temperature per location (lat, lng) for each day in range between first visit in that location and last visit recorded.

Will get the data sequantialy for each property. This can be slow but for now we have only 80 properties, so it's not a pain yet. An option would be using `async requests`.

In [44]:
properties_visit_dates_range.head(2)

property_id                begin_date                  end_date  \
0            1 2022-01-10 12:00:00-03:00 2022-01-28 14:00:00-03:00   
1            2 2022-01-20 13:00:00-03:00 2022-01-20 15:00:00-03:00   

        dates_diff  
0 18 days 02:00:00  
1  0 days 02:00:00

In [45]:
properties_data.head(2)

property_id    type_house  business_type  bedrooms  bathrooms  \
0            1  departamento         Rental         1          1   
1            2  departamento  Rental & Sale         1          1   

   parking_lots  services  balcony   pool  latitude  longitude    localidad  \
0             1       3.0        0  False  4.870956  -74.05804        Chí­A   
1             2       NaN        1  False  4.623068  -74.07403  Teusaquillo   

           city                  region   country  
0  Cundinamarca  Región De Cundinamarca  Colombia  
1        Bogotá  Región De Cundinamarca  Colombia

In [46]:
# adding lat and lng to properties_visit_dates_range
properties_visit_dates_range = pd.merge(properties_visit_dates_range, properties_data[['property_id', 'latitude', 'longitude']], on='property_id')
properties_visit_dates_range.head(2)

property_id                begin_date                  end_date  \
0            1 2022-01-10 12:00:00-03:00 2022-01-28 14:00:00-03:00   
1            2 2022-01-20 13:00:00-03:00 2022-01-20 15:00:00-03:00   

        dates_diff  latitude  longitude  
0 18 days 02:00:00  4.870956  -74.05804  
1  0 days 02:00:00  4.623068  -74.07403

In [47]:
properties_visit_dates_range.shape

(80, 6)

In [48]:
def get_weather_data(lat: float, lng: float, start_date: datetime, end_date: datetime, api_key:str) -> dict:
    """ Calls weather API and get weather data for a given lat, lng, start_date, end_date
    
    Args:
        lat (float): latitude
        lng (float): longitude
        start_date (datetime): start date
        end_date (datetime): end date
        api_key (str): API key
    
    Returns:
        dict: weather data
    """
    BASE_URL = "https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/{lat},{lng}/{start_date}/{end_date}?key={API_KEY}&include=days"
    url = BASE_URL.format(lat=lat, lng=lng, start_date=f"{start_date:%Y-%m-%d}", end_date=f"{end_date:%Y-%m-%d}", API_KEY=api_key)
    r = requests.get(url=url)
    if not r.ok:
        print("Error calling weather API for lat: {}, lng: {}, start_date: {}, end_date: {}".format(lat, lng, start_date, end_date))
        print(r.text)
        return None
    return r.json()
    

In [49]:
# testing fucntion
res = get_weather_data(lat=test_property.latitude, lng=test_property.longitude, start_date=min_date, end_date=max_date, api_key=os.getenv('WEATHER_API_KEY'))
res

{'queryCost': 12,
 'latitude': 4.7239094,
 'longitude': -74.042336,
 'resolvedAddress': '4.7239094,-74.042336',
 'address': '4.7239094,-74.042336',
 'timezone': 'America/Bogota',
 'tzoffset': -5.0,
 'days': [{'datetime': '2022-01-05',
   'datetimeEpoch': 1641358800,
   'tempmax': 71.8,
   'tempmin': 50.8,
   'temp': 59.4,
   'feelslikemax': 71.8,
   'feelslikemin': 50.8,
   'feelslike': 59.4,
   'dew': 49.4,
   'humidity': 72.99,
   'precip': 0.0,
   'precipprob': None,
   'precipcover': 0.0,
   'preciptype': None,
   'snow': None,
   'snowdepth': None,
   'windgust': None,
   'windspeed': 13.8,
   'winddir': 192.6,
   'pressure': 1023.3,
   'cloudcover': 56.5,
   'visibility': 6.2,
   'solarradiation': None,
   'solarenergy': None,
   'uvindex': None,
   'sunrise': '06:05:56',
   'sunriseEpoch': 1641380756,
   'sunset': '17:57:27',
   'sunsetEpoch': 1641423447,
   'moonphase': 0.05,
   'conditions': 'Partially cloudy',
   'description': 'Partly cloudy throughout the day.',
   'icon': 

In [50]:
def convert_weather_data_to_df(weather_data: dict) -> pd.DataFrame:
    """ Converts weather data to dataframe
    
    Args:
        weather_data (dict): weather data
    
    Returns:
        pd.DataFrame: weather dataframe
    """
    assert "days" in weather_data.keys(), "weather_data must contain 'days' key"

    weather_data_mask = pd.DataFrame()
    for day in weather_data["days"]:
        day_mask = copy.deepcopy(day)
        day_mask.pop("stations")
        res_df = pd.DataFrame([day_mask])
        weather_data_mask = pd.concat([weather_data_mask, res_df], ignore_index=True)
    assert weather_data_mask.groupby('datetime').count().max().iloc[0] == 1, "weather_data_mask must contain unique datetime"

    weather_data_mask.insert(0, 'latitude', weather_data['latitude'])
    weather_data_mask.insert(0, 'longitude', weather_data['longitude'])

    return weather_data_mask

In [51]:
# testing function
convert_weather_data_to_df(weather_data=res).head(2)

longitude  latitude    datetime  datetimeEpoch  tempmax  tempmin  temp  \
0 -74.042336  4.723909  2022-01-05     1641358800     71.8     50.8  59.4   
1 -74.042336  4.723909  2022-01-06     1641445200     71.8     48.2  57.1   

  feelslikemax feelslikemin feelslike  ...   sunrise  sunriseEpoch    sunset  \
0         71.8         50.8      59.4  ...  06:05:56    1641380756  17:57:27   
1         71.8         45.6      56.9  ...  06:06:20    1641467180  17:57:56   

  sunsetEpoch moonphase        conditions                        description  \
0  1641423447      0.05  Partially cloudy  Partly cloudy throughout the day.   
1  1641509876      0.09  Partially cloudy  Partly cloudy throughout the day.   

                icon source  severerisk  
0  partly-cloudy-day    obs         NaN  
1  partly-cloudy-day    obs         NaN  

[2 rows x 37 columns]

In [59]:
# main extraction logic
failed_requests = []
weather_data = pd.DataFrame()
total_lenght = len(properties_visit_dates_range)
for idx, row in properties_visit_dates_range.iterrows():
    print(f"Getting data {idx+1}/{total_lenght}...")
    weather_dict = get_weather_data(lat=row.latitude, lng=row.longitude, start_date=row.begin_date, end_date=row.end_date, api_key=os.getenv('WEATHER_API_KEY'))
    
    # in case it failed
    if weather_dict is None:
        failed_requests.append(row.to_dict())
        continue

    if "days" not in weather_dict.keys():
        print("An unkonw error ocuured for property_id: {}".format(row.property_id))
        failed_requests.append(row.property_id)
        print("request returned:", weather_dict)
        continue

    weather_df_mask = convert_weather_data_to_df(weather_data=weather_dict)
    weather_df_mask.insert(0, "property_id", row.property_id)
    weather_data = pd.concat([weather_data, weather_df_mask], ignore_index=True)
weather_data 

Getting data 1/80...
Getting data 2/80...
Getting data 3/80...
Getting data 4/80...
Getting data 5/80...
Getting data 6/80...
Getting data 7/80...
Getting data 8/80...
Getting data 9/80...
Getting data 10/80...
Getting data 11/80...
Getting data 12/80...
Getting data 13/80...
Getting data 14/80...
Getting data 15/80...
Getting data 16/80...
Getting data 17/80...
Getting data 18/80...
Getting data 19/80...
Getting data 20/80...
Getting data 21/80...
Getting data 22/80...
Getting data 23/80...
Getting data 24/80...
Getting data 25/80...
Getting data 26/80...
Getting data 27/80...
Getting data 28/80...
Getting data 29/80...
Getting data 30/80...
Getting data 31/80...
Getting data 32/80...
Getting data 33/80...
Getting data 34/80...
Getting data 35/80...
Getting data 36/80...
Getting data 37/80...
Getting data 38/80...
Getting data 39/80...
Getting data 40/80...
Getting data 41/80...
Getting data 42/80...
Getting data 43/80...
Getting data 44/80...
Getting data 45/80...
Getting data 46/80.

property_id  longitude  latitude    datetime  datetimeEpoch  tempmax  \
0              1 -74.058040  4.870956  2022-01-10     1641790800     66.0   
1              1 -74.058040  4.870956  2022-01-11     1641877200     66.2   
2              1 -74.058040  4.870956  2022-01-12     1641963600     64.5   
3              1 -74.058040  4.870956  2022-01-13     1642050000     62.6   
4              1 -74.058040  4.870956  2022-01-14     1642136400     65.7   
..           ...        ...       ...         ...            ...      ...   
717           79 -74.226036  4.725887  2022-02-27     1645938000     64.4   
718           79 -74.226036  4.725887  2022-02-28     1646024400     59.3   
719           80 -74.127815  4.587935  2022-02-26     1645851600     65.0   
720           80 -74.127815  4.587935  2022-02-27     1645938000     64.8   
721           80 -74.127815  4.587935  2022-02-28     1646024400     59.9   

     tempmin  temp feelslikemax feelslikemin  ... severerisk   sunrise  \
0       46.2  55.8         66.0         44.2  ...       10.0  06:08:07   
1       50.4  57.7         66.2         50.4  ...       10.0  06:08:27   
2       50.0  57.4         64.5         50.0  ...       30.0  06:08:47   
3       50.0  55.1         62.6         50.0  ...       10.0  06:09:06   
4       48.2  56.3         65.7         47.3  ...       10.0  06:09:24   
..       ...   ...          ...          ...  ...        ...       ...   
717     53.6  57.0         64.4         53.6  ...       10.0  06:08:58   
718     53.4  55.8         59.3         53.4  ...       10.0  06:08:39   
719     52.1  58.3         65.0         52.1  ...       10.0  06:08:47   
720     53.6  57.2         64.8         53.6  ...       10.0  06:08:29   
721     53.4  56.2         59.9         53.4  ...       10.0  06:08:11   

     sunriseEpoch    sunset sunsetEpoch moonphase              conditions  \
0      1641812887  17:59:38  1641855578      0.28  Rain, Partially cloudy   
1      1641899307  18:00:05  1641942005      0.33  Rain, Partially cloudy   
2      1641985727  18:00:32  1642028432      0.38  Rain, Partially cloudy   
3      1642072146  18:00:58  1642114858      0.41  Rain, Partially cloudy   
4      1642158564  18:01:24  1642201284      0.45  Rain, Partially cloudy   
..            ...       ...         ...       ...                     ...   
717    1645960138  18:10:17  1646003417      0.94  Rain, Partially cloudy   
718    1646046519  18:10:13  1646089813      0.97  Rain, Partially cloudy   
719    1645873727  18:10:02  1645917002      0.89  Rain, Partially cloudy   
720    1645960109  18:09:58  1646003398      0.94  Rain, Partially cloudy   
721    1646046491  18:09:54  1646089794      0.97  Rain, Partially cloudy   

                                           description  icon source  
0          Partly cloudy throughout the day with rain.  rain    obs  
1          Partly cloudy throughout the day with rain.  rain    obs  
2    Partly cloudy throughout the day with storms p...  rain    obs  
3          Partly cloudy throughout the day with rain.  rain    obs  
4          Partly cloudy throughout the day with rain.  rain    obs  
..                                                 ...   ...    ...  
717        Partly cloudy throughout the day with rain.  rain    obs  
718  Partly cloudy throughout the day with a chance...  rain    obs  
719  Partly cloudy throughout the day with a chance...  rain    obs  
720        Partly cloudy throughout the day with rain.  rain    obs  
721  Partly cloudy throughout the day with a chance...  rain    obs  

[722 rows x 38 columns]

In [62]:
len(failed_requests)

0

In [64]:
if len(weather_data.property_id.unique()) == len(properties_data.property_id.unique()):
    print("All data properly retrieved")

All data properly retrieved


In [65]:
# Will save the data because should not have a lot of credits left to re-run the extraction of data.
weather_data.to_csv('./data/weather_data.csv', index=False)

In [55]:
# menos mal... obviamente windows se cayó en la mitad...
weather_data= pd.read_csv('./data/weather_data.csv')
weather_data["datetime"] = pd.to_datetime(weather_data["datetime"])
weather_data.head(2)

property_id  longitude  latitude   datetime  datetimeEpoch  tempmax  \
0            1  -74.05804  4.870956 2022-01-10     1641790800     66.0   
1            1  -74.05804  4.870956 2022-01-11     1641877200     66.2   

   tempmin  temp  feelslikemax  feelslikemin  ...  severerisk   sunrise  \
0     46.2  55.8          66.0          44.2  ...        10.0  06:08:07   
1     50.4  57.7          66.2          50.4  ...        10.0  06:08:27   

   sunriseEpoch    sunset  sunsetEpoch  moonphase              conditions  \
0    1641812887  17:59:38   1641855578       0.28  Rain, Partially cloudy   
1    1641899307  18:00:05   1641942005       0.33  Rain, Partially cloudy   

                                   description  icon  source  
0  Partly cloudy throughout the day with rain.  rain     obs  
1  Partly cloudy throughout the day with rain.  rain     obs  

[2 rows x 38 columns]

------

## Q3. ¿Cuál era la temperatura promedio de todas las visitas que realizó el usuario con ID 2?

In [67]:
visits_data.describe()

scheduled_id  property_id
count     425.000000   425.000000
mean   188094.894118    40.087059
std     13820.402675    22.943521
min    160130.000000     1.000000
25%    176671.000000    21.000000
50%    190273.000000    38.000000
75%    200307.000000    61.000000
max    208926.000000    80.000000

Según lo que entiendo de la documentación de la tarea, un usuario es un propetario y no un cliente que hace visitas.

>users.csv: Contiene la información de los propietarios y su relación con la propiedad. Las columnas de este archivo son las siguientes: property_id, user_id, name, last_name & country. 

**Voy a asumir que se refiere a todas las visitas efectivas que se le hizo a la propiedad del usuario con ID 2**

Se calculará el promedio sobre el promedio diario de temperatura. Esto puede ser missleading respecto a la data real del día ya que el promedio esconde harto. Pero es una estimación en base a la limitancia de la API de VisualCrossing. 

Una alternativa, ya que se tiene la temperatura maxima, minima y promedio por día, es asumir que se distribuyen normal y extrapolar la temperatura de la hora en base a esos 3 datos mencionados. 

In [56]:
users_data.loc[users_data['user_id']==2]

property_id  user_id first_name last_name               address
1           34        2        Art    Venere  8 W Cerritos Ave #54

In [57]:
p = users_data.loc[users_data['user_id']==2].iloc[0]

In [68]:
visits = visits_data[visits_data['property_id']==p.property_id].reset_index(drop=True)
visits

scheduled_id  property_id                begin_date  \
0        186033           34 2022-01-29 12:00:00-03:00   
1        187080           34 2022-01-28 11:00:00-03:00   
2        186857           34 2022-01-28 11:30:00-03:00   
3        180698           34 2022-01-22 20:00:00-03:00   
4        177506           34 2022-01-23 13:00:00-03:00   
5        178172           34 2022-01-21 18:00:00-03:00   
6        177505           34 2022-01-23 09:00:00-03:00   

                   end_date type_visit     status  
0 2022-01-29 14:00:00-03:00      Visit  Cancelled  
1 2022-01-29 01:00:00-03:00      Visit       Done  
2 2022-01-28 14:00:00-03:00      Visit       Done  
3 2022-01-22 22:00:00-03:00      Visit  Cancelled  
4 2022-01-23 15:00:00-03:00      Visit  Cancelled  
5 2022-01-21 20:00:00-03:00      Visit  Cancelled  
6 2022-01-23 11:00:00-03:00      Visit  Cancelled

In [62]:
weather_data.head(1)

property_id  longitude  latitude   datetime  datetimeEpoch  tempmax  \
0            1  -74.05804  4.870956 2022-01-10     1641790800     66.0   

   tempmin  temp  feelslikemax  feelslikemin  ...  severerisk   sunrise  \
0     46.2  55.8          66.0          44.2  ...        10.0  06:08:07   

   sunriseEpoch    sunset  sunsetEpoch  moonphase              conditions  \
0    1641812887  17:59:38   1641855578       0.28  Rain, Partially cloudy   

                                   description  icon  source  
0  Partly cloudy throughout the day with rain.  rain     obs  

[1 rows x 38 columns]

In [76]:
visits['datetime'] = visits['begin_date'].dt.date
visits['datetime'] = pd.to_datetime(visits['datetime'])

In [81]:
# Merging visits and weather data so we can have full detail of the weather of each visit
visits_with_weather = visits.merge(weather_data, on=['datetime', 'property_id'], how='inner')
visits_with_weather.head(2)

scheduled_id  property_id                begin_date  \
0        186033           34 2022-01-29 12:00:00-03:00   
1        187080           34 2022-01-28 11:00:00-03:00   

                   end_date type_visit     status   datetime  longitude  \
0 2022-01-29 14:00:00-03:00      Visit  Cancelled 2022-01-29   -74.0735   
1 2022-01-29 01:00:00-03:00      Visit       Done 2022-01-28   -74.0735   

   latitude  datetimeEpoch  ...  severerisk   sunrise  sunriseEpoch    sunset  \
0  4.618363     1643432400  ...        10.0  06:11:56    1643454716  18:07:04   
1  4.618363     1643346000  ...        30.0  06:11:51    1643368311  18:06:48   

   sunsetEpoch  moonphase              conditions  \
0   1643497624       0.95  Rain, Partially cloudy   
1   1643411208       0.91  Rain, Partially cloudy   

                                         description  icon  source  
0  Partly cloudy throughout the day with late aft...  rain     obs  
1  Partly cloudy throughout the day with storms p...  rain     obs  

[2 rows x 43 columns]

In [82]:
# Esuring we got oficially all the needed dates for each visit
assert visits.shape[0] == visits_with_weather.shape[0], "duoplicated!!!"

In [96]:
visits_with_weather[['status', 'temp']].groupby('status').mean()

temp
status         
Cancelled  58.1
Done       58.3

### Solution Q3
Temperatura promedio para las visitas realizadas para las propiedades del usuario con ID 2 es de **58.3** grados Fahrenheit.

## Q4. ¿Cuál es la temperatura promedio de las visitas para los días con lluvia?

Se define como día con lluvia como día que haya caido al menos una gota de agua, es decir, precipitación en el día mayor a 0

In [85]:
weather_data.head(2)

property_id  longitude  latitude   datetime  datetimeEpoch  tempmax  \
0            1  -74.05804  4.870956 2022-01-10     1641790800     66.0   
1            1  -74.05804  4.870956 2022-01-11     1641877200     66.2   

   tempmin  temp  feelslikemax  feelslikemin  ...  severerisk   sunrise  \
0     46.2  55.8          66.0          44.2  ...        10.0  06:08:07   
1     50.4  57.7          66.2          50.4  ...        10.0  06:08:27   

   sunriseEpoch    sunset  sunsetEpoch  moonphase              conditions  \
0    1641812887  17:59:38   1641855578       0.28  Rain, Partially cloudy   
1    1641899307  18:00:05   1641942005       0.33  Rain, Partially cloudy   

                                   description  icon  source  
0  Partly cloudy throughout the day with rain.  rain     obs  
1  Partly cloudy throughout the day with rain.  rain     obs  

[2 rows x 38 columns]

In [89]:
visits_data['datetime'] = visits_data['begin_date'].dt.date
visits_data['datetime'] = pd.to_datetime(visits_data['datetime'])
visits_data.head(2)

scheduled_id  property_id                begin_date  \
0        169548            1 2022-01-13 10:00:00-03:00   
1        184763            1 2022-01-26 18:00:00-03:00   

                   end_date type_visit     status   datetime  
0 2022-01-13 12:00:00-03:00      Visit  Cancelled 2022-01-13  
1 2022-01-26 20:00:00-03:00      Visit  Cancelled 2022-01-26

In [92]:
all_visits_with_weather = visits_data.merge(weather_data, on=['datetime', 'property_id'], how='inner')
assert all_visits_with_weather.shape[0] == visits_data.shape[0], "duoplicated!!!"
all_visits_with_weather.head(2)

scheduled_id  property_id                begin_date  \
0        169548            1 2022-01-13 10:00:00-03:00   
1        174344            1 2022-01-13 15:00:00-03:00   

                   end_date type_visit     status   datetime  longitude  \
0 2022-01-13 12:00:00-03:00      Visit  Cancelled 2022-01-13  -74.05804   
1 2022-01-13 17:00:00-03:00      Visit  Cancelled 2022-01-13  -74.05804   

   latitude  datetimeEpoch  ...  severerisk   sunrise  sunriseEpoch    sunset  \
0  4.870956     1642050000  ...        10.0  06:09:06    1642072146  18:00:58   
1  4.870956     1642050000  ...        10.0  06:09:06    1642072146  18:00:58   

   sunsetEpoch  moonphase              conditions  \
0   1642114858       0.41  Rain, Partially cloudy   
1   1642114858       0.41  Rain, Partially cloudy   

                                   description  icon  source  
0  Partly cloudy throughout the day with rain.  rain     obs  
1  Partly cloudy throughout the day with rain.  rain     obs  

[2 rows x 43 columns]

In [86]:
weather_data.columns

Index(['property_id', 'longitude', 'latitude', 'datetime', 'datetimeEpoch',
       'tempmax', 'tempmin', 'temp', 'feelslikemax', 'feelslikemin',
       'feelslike', 'dew', 'humidity', 'precip', 'precipprob', 'precipcover',
       'preciptype', 'snow', 'snowdepth', 'windgust', 'windspeed', 'winddir',
       'pressure', 'cloudcover', 'visibility', 'solarradiation', 'solarenergy',
       'uvindex', 'severerisk', 'sunrise', 'sunriseEpoch', 'sunset',
       'sunsetEpoch', 'moonphase', 'conditions', 'description', 'icon',
       'source'],
      dtype='object')

In [100]:
x = all_visits_with_weather.loc[all_visits_with_weather['precip']>0, :]
x[['status', 'temp']].groupby('status').mean()

temp
status              
Cancelled  57.154297
Done       57.630328

In [101]:
all_visits_with_weather.loc[all_visits_with_weather['precip']>0, 'temp'].mean()

57.30793650793651

### Solucion Q4
Temperatura promedio para los días con lluvia es **57.30** grados Fahrenheit. <br>
Las visitas efectivas tuvieron una temperatura promedio de **57.63** grados Fahrenheit.

## Q5. ¿Cuál es la temperatura promedio para las visitas realizadas en la localidad de Suba?

Asumo que cuando se refiere a visitas realizadas se refiere a visitas no canceladas

In [104]:
all_visits_with_weather.columns

Index(['scheduled_id', 'property_id', 'begin_date', 'end_date', 'type_visit',
       'status', 'datetime', 'longitude', 'latitude', 'datetimeEpoch',
       'tempmax', 'tempmin', 'temp', 'feelslikemax', 'feelslikemin',
       'feelslike', 'dew', 'humidity', 'precip', 'precipprob', 'precipcover',
       'preciptype', 'snow', 'snowdepth', 'windgust', 'windspeed', 'winddir',
       'pressure', 'cloudcover', 'visibility', 'solarradiation', 'solarenergy',
       'uvindex', 'severerisk', 'sunrise', 'sunriseEpoch', 'sunset',
       'sunsetEpoch', 'moonphase', 'conditions', 'description', 'icon',
       'source'],
      dtype='object')

In [105]:
properties_data.columns

Index(['property_id', 'type_house', 'business_type', 'bedrooms', 'bathrooms',
       'parking_lots', 'services', 'balcony', 'pool', 'latitude', 'longitude',
       'localidad', 'city', 'region', 'country'],
      dtype='object')

In [107]:
all_visits_with_weather_and_location = all_visits_with_weather.merge(properties_data[['property_id', 'localidad']], on=['property_id'], how='inner')
assert all_visits_with_weather_and_location.shape[0] == visits_data.shape[0], "Duplicated!!"
all_visits_with_weather_and_location.head(2)

scheduled_id  property_id                begin_date  \
0        169548            1 2022-01-13 10:00:00-03:00   
1        174344            1 2022-01-13 15:00:00-03:00   

                   end_date type_visit     status   datetime  longitude  \
0 2022-01-13 12:00:00-03:00      Visit  Cancelled 2022-01-13  -74.05804   
1 2022-01-13 17:00:00-03:00      Visit  Cancelled 2022-01-13  -74.05804   

   latitude  datetimeEpoch  ...   sunrise  sunriseEpoch    sunset  \
0  4.870956     1642050000  ...  06:09:06    1642072146  18:00:58   
1  4.870956     1642050000  ...  06:09:06    1642072146  18:00:58   

   sunsetEpoch  moonphase              conditions  \
0   1642114858       0.41  Rain, Partially cloudy   
1   1642114858       0.41  Rain, Partially cloudy   

                                   description  icon  source  localidad  
0  Partly cloudy throughout the day with rain.  rain     obs      Chí­A  
1  Partly cloudy throughout the day with rain.  rain     obs      Chí­A  

[2 rows x 44 columns]

In [112]:
suba_visits = all_visits_with_weather_and_location[all_visits_with_weather_and_location['localidad']=='Suba']
suba_visits['temp'].mean()

57.589534883720916

In [114]:
suba_visits[['status', 'temp']].groupby('status').mean()

temp
status              
Cancelled  57.713725
Done       57.408571

### Solucion Q5
Temperatura promedio para visitas realizadas (status = "Done") es de **57.40** grados Fahrenheit. <br>
Promedio general de temperatura de las visitas agendadas es de **57.58** grados Fahrenheit. 